modules/core/04-channel/keeper/upgrade.go

```diff
@@ -86,17 +88,249 @@ func (k Keeper) WriteUpgradeInitChannel(
+		if counterpartyUpgradeSequence > upgradeSequence {
+			upgradeSequence = counterpartyUpgradeSequence
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+		} else {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			// the upgrade sequence is incremented so both sides start the next upgrade with a fresh sequence.
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+
+			// TODO: emit error receipt events
+
+			// do we want to return upgrade sequence here to include in response??
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", errorReceipt.GetSequence())
+		}
+
+		// this is first message in upgrade handshake on this chain so we must store original channel in restore channel path
+		// in case we need to restore channel later.
+		k.SetUpgradeRestoreChannel(ctx, portID, channelID, channel)
+	case types.INITUPGRADE:
+		upgradeSequence, found = k.GetUpgradeSequence(ctx, portID, channelID)
+		if !found {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		if upgradeSequence != counterpartyUpgradeSequence {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade sequence %d was not equal to the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+
+			// set to the max of the two
+			if counterpartyUpgradeSequence > upgradeSequence {
+				upgradeSequence = counterpartyUpgradeSequence
+				k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+			}
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		// if there is a crossing hello, i.e an UpgradeInit has been called on both channelEnds,
+		// then we must ensure that the proposedUpgrade by the counterparty is the same as the currentChannel
+		// except for the channel state (upgrade channel will be in TRYUPGRADE and current channel will be in INITUPGRADE)
+		// if the proposed upgrades on either side are incompatible, then we will restore the channel and cancel the upgrade.
+		channel.State = types.TRYUPGRADE
+		k.SetChannel(ctx, portID, channelID, channel)
```

Comment:
Should this `SetChannel` (maybe for clarity) be done at the end?

Reply:
The updated channel state is used just a few lines down and I think separating them out makes things a little harder to follow in this case

Reply:
I tend to agree with @chatton that it is clearer to keep the logic together, rather than having the channel state being updated in one place and then actually set further down with the potential restore channel logic happening in between